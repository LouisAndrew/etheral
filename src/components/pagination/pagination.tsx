import React from 'react';

import { Flex, Text, Box } from 'rebass';
import { Icon } from '@iconify/react';
import speedMiniFill from '@iconify/icons-ri/speed-mini-fill';
import rewindMiniFill from '@iconify/icons-ri/rewind-mini-fill';

/**
 * @param numOfPages: number of pages to be rendered.
 * @param current: num of current in-view index.
 * @param handleClickPage: function to be called when a page number is clicked.
 */
type Props = {
    numOfPages: number;
    current: number;
    handleClickPage: (pageNum: number) => void;
    goToFirst: () => void;
    goToLast: () => void;
};

const createPagination = (
    first: number,
    last: number,
    numOfPages: number
): number[] => {
    const MAX_PAGE_COUNT = 5; // max num of pages to display on the pagination component.

    const pageCount = last - first;

    if (pageCount < MAX_PAGE_COUNT) {
        if (MAX_PAGE_COUNT > numOfPages) {
            if (last === numOfPages) {
                const firstFromLast = last - numOfPages;
                return [...new Array(numOfPages)].map(
                    (_, i) => firstFromLast + i
                );
            } else {
                return [...new Array(numOfPages)].map((_, i) => first + i);
            }
        }

        if (first === 0) {
            // return 0 to 5.
            return [...new Array(MAX_PAGE_COUNT)].map((_, i) => first + i);
        } else if (last === numOfPages) {
            const firstFromLast = last - MAX_PAGE_COUNT;
            return [...new Array(MAX_PAGE_COUNT)].map(
                (_, i) => firstFromLast + i
            );
        } else {
            return [...new Array(pageCount + 1)].map((_, i) => first + i); // +1 -> including the current page.
        }
    } else {
        // no probs here
        return [...new Array(pageCount + 1)].map((_, i) => first + i); // +1 -> including the current page.
    }
};

const Pagination: React.FC<Props> = ({
    numOfPages,
    current,
    handleClickPage,
    goToFirst,
    goToLast,
}) => {
    const firstPage = current - 2 < 0 ? 0 : current - 2;
    const lastPage = current + 2 >= numOfPages ? numOfPages : current + 2;

    const pagination = createPagination(firstPage, lastPage, numOfPages);

    const handleClick = (num: number) => {
        if (window) {
            window.scrollTo(0, 0);
        }

        handleClickPage(num);
    };

    const buttonStyles = {
        height: '100%',
        position: 'relative',
        top: '50%',
        transition: '0.2s',
        '&:hover': {
            cursor: 'pointer',
            transform: 'scale(1.2) !important',
        },
    };

    return (
        <Flex as="nav" alignItems="center">
            {!pagination.includes(0) && (
                <Flex
                    id="go-to-first"
                    role="button"
                    onClick={goToFirst}
                    variant="center"
                    sx={{ svg: buttonStyles }}
                >
                    <Icon icon={rewindMiniFill} />
                </Flex>
            )}
            {pagination.map(num => (
                <Text
                    key={`pagination-${num}`}
                    onClick={() => handleClick(num)}
                    role="listitem"
                    fontWeight={num === current ? 'bold' : 300}
                    fontFamily="body"
                    fontSize={[1, 1, 2]}
                    color="black.0"
                    height={[24]}
                    width={[24]}
                    mx={[3]}
                    sx={{
                        display: 'flex',
                        alignItems: 'center',
                        justifyContent: 'center',
                        '&:hover': { cursor: 'pointer' },
                    }}
                >
                    {num + 1}
                </Text>
            ))}
            {!pagination.includes(numOfPages - 1) && (
                <Flex
                    id="go-to-last"
                    role="button"
                    onClick={goToLast}
                    variant="center"
                    sx={{ svg: buttonStyles }}
                >
                    <Icon icon={speedMiniFill} />
                </Flex>
            )}
        </Flex>
    );
};

export { Pagination };
