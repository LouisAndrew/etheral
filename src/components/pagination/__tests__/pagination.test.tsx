import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import userEvent from '@testing-library/user-event';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import { Pagination } from '../pagination';

describe('Pagination', () => {
    const mockCurrent = 2;
    const mockNumOfPages = 5;
    const mockClickPage = jest.fn(() => {});

    const Element = (
        <Pagination
            current={mockCurrent}
            numOfPages={mockNumOfPages}
            handleClickPage={mockClickPage}
        />
    );

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(Element, div);
    });

    it('Should render all page links correctly', () => {
        const { queryByText } = render(Element);

        for (let i = 0; i < mockNumOfPages; i++) {
            const PageElement = queryByText((i + 1).toString());

            if (PageElement) {
                expect(PageElement).toBeInTheDocument();
            } else {
                fail();
            }
        }
    });

    // provide rebass container?

    // it('Should render element with virtual hierarchy on current page number', () => {
    //     const { getByText } = render(Element);
    //     const CurrentNum = getByText((mockCurrent + 1).toString());

    //     expect(CurrentNum).toHaveStyle(
    //         `background-color: ${theme.colors.black[0]};`
    //     );
    // });

    it('Should calls provided function when a page num is clicked', () => {
        const { getByText } = render(Element);
        const FirstPage = getByText((1).toString());

        userEvent.click(FirstPage);
        expect(mockClickPage).toBeCalledWith(0);
    });

    it('matches snapshot', () => {
        const run = true;

        if (run) {
            const tree = renderer.create(Element).toJSON();
            expect(tree).toMatchSnapshot();
        }
    });
});
