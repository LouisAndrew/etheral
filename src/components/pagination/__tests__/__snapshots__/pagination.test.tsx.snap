// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination matches snapshot 1`] = `
<nav
  className="css-4cffwv"
>
  <div
    className="css-1u57j0c"
    onClick={[Function]}
    role="listitem"
  >
    1
  </div>
  <div
    className="css-1u57j0c"
    onClick={[Function]}
    role="listitem"
  >
    2
  </div>
  <div
    className="css-154jwdi"
    onClick={[Function]}
    role="listitem"
  >
    3
  </div>
  <div
    className="css-1u57j0c"
    onClick={[Function]}
    role="listitem"
  >
    4
  </div>
  <div
    className="css-1u57j0c"
    onClick={[Function]}
    role="listitem"
  >
    5
  </div>
</nav>
`;
